api/authenticators/onelogin_duo/authenticator.go

```diff
@@ -1,102 +1,72 @@
 	"github.com/rnikoopour/onelogin"
-	"github.com/sirupsen/logrus"
 )
 
-type OneLoginSaml struct {
-	b64String    string
-	samlResponse *saml.Response
-	logger       *logrus.Entry
-}
-
-func (s OneLoginSaml) GetBase64String() string {
-	return s.b64String
-}
-
-func (s OneLoginSaml) GetSamlResponse() *saml.Response {
-	return s.samlResponse
-}
-
-type DuoMFA struct {
-	logger *logrus.Entry
+type Authenticator struct {
+	settings *settings.Settings
+	mfa      duo.Duo
 }
 
-func NewDuoMFA(logger *logrus.Entry) *DuoMFA {
-	return &DuoMFA{
-		logger: logger}
-}
-
-func (d *DuoMFA) Do(args ...string) (string, error) {
-	d.logger.Debug("prepping push")
-	duo := NewDuo(d.logger)
-
-	if len(args) < 4 {
-		return "", ErrorDuoArgsError
+// New creates a new OneLogin authenticator with the given MFA instance
+func New(settings *settings.Settings, mfa duo.Duo) *Authenticator {
+	return &Authenticator{
+		settings: settings,
+		mfa:      mfa,
 	}
-
-	txSig := args[0]
-	stateToken := args[1]
-	callbackUrl := args[2]
-	apiHost := args[3]
-
-	d.logger.Info("sending duo push")
-	return duo.SendPush(txSig, stateToken, callbackUrl, apiHost)
-}
-
-type OneLoginAuthenticator struct {
-	Settings *settings.Settings
-	MFA      authenticators.MFA
-	logger   *logrus.Entry
 }
 
-func New(logger *logrus.Entry, settings *settings.Settings) authenticators.Authenticator {
-	return &OneLoginAuthenticator{
-		Settings: settings,
-		logger:   logger}
-}
+func (a *Authenticator) Authenticate(ctx context.Context, creds core.Credentials) (core.User, error) {
+	o := newOneLogin(a.settings)
+	user, err := o.SamlClient.Oauth.Authenticate(ctx, creds.Username, creds.Password)
+	if err != nil {
+		return core.User{}, err
+	}
 
-func (ola *OneLoginAuthenticator) SetMFA(mfa authenticators.MFA) {
-	ola.MFA = mfa
+	return core.User{ID: strconv.FormatInt(user.ID, 10)}, nil
 }
 
-func (ola *OneLoginAuthenticator) Authenticate(username string, password string) ([]authenticators.Account, error) {
-	oneLoginClient := NewOneLogin(ola.Settings, ola.logger)
-
-	authenticatedUser, err := oneLoginClient.AuthenticateUser(username, password)
+func (a *Authenticator) ListApplications(ctx context.Context, user core.User) ([]core.Application, error) {
+	o := newOneLogin(a.settings)
+	id, err := strconv.ParseInt(user.ID, 10, 64)
 	if err != nil {
-		ola.logger.Error("failed to authenticate user reason: ", err.Error())
 		return nil, err
 	}
 
-	allUserApps, err := oneLoginClient.GetUserApps(authenticatedUser)
+	apps, err := o.ReadUserClient.User.GetApps(ctx, id)
 	if err != nil {
-		ola.logger.Error("unable to get user apps reason: ", err.Error())
 		return nil, err
 	}
 
-	accounts := make([]authenticators.Account, len(allUserApps))
-	for index, app := range allUserApps {
-		accounts[index] = app
+	applications := []core.Application{}
+	for _, app := range *apps {
+		if strings.HasPrefix(app.Name, "AWS") {
+			applications = append(applications, core.Application{
+				LegacyID: uint(app.ID),
```

Comment:
`LegacyID` is used to facilitate support for old clients - this is listed inline. The reason for this is just for JSON Marshalling - `LegacyID` is marshalled to `id` which old clients read and must be a number. Unfortunately, Okta applications have string identifiers, so we can't use the same property if we want to support backwards compatibility, so `ID` is the string-formatted ID for newer clients that want to continue using onelogin